Fix bugs in metrics calculation in inference benchmark and rerun baseline

mikaylagawarecki opened the pull request:
Before this PR, each `request_time` was separated by the time for a `torch.randn(...)` to create the fake `data` tensor on CPU. This meant that the gap between `request_times` **scaled with the batch_size**. So the latency comparisons across batch sizes were inaccurate. In this PR we generate all the fake data outside the loop to avoid this.

Other bug fixes:
- Only start polling GPU utilization after warmup event is complete
- Correct calculation of throughput: previously `(num_batches * batch_size) / sum(response_times)`, should have been `(num_batches * batch_size) / (last_response_time - first_request_time)`
- Make sure that response sent back to frontend is on CPU
- Use a lock to ensure writing to `metrics_dict` in `metrics_thread` and `gpu_utilization_thread` in a thread-safe manner

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116190
* #116189
* __->__ #116188
* #116187
* #115286

mikaylagawarecki committed:
[easy] Generate all requests before putting first request into queue in inference benchmark

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix bugs in metrics calculation in inference benchmark and rerun baseline"

Before this PR, each `request_time` was separated by the time for a `torch.randn(...)` to create the fake `data` tensor on CPU. This meant that the gap between `request_times` **scaled with the batch_size**. So the latency comparisons across batch sizes were inaccurate. In this PR we generate all the fake data outside the loop to avoid this.

Other bug fixes:
- Only start polling GPU utilization after warmup event is complete
- Correct calculation of throughput: previously `(num_batches * batch_size) / sum(response_times)`, should have been `(num_batches * batch_size) / (last_response_time - first_request_time)`
- Make sure that response sent back to frontend is on CPU
- Use a semaphore to ensure writing to `metrics_dict` in `metrics_thread` and `gpu_utilization_thread` in a thread-safe manner




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix bugs in metrics calculation in inference benchmark and rerun baseline"

Before this PR, each `request_time` was separated by the time for a `torch.randn(...)` to create the fake `data` tensor on CPU. This meant that the gap between `request_times` **scaled with the batch_size**. So the latency comparisons across batch sizes were inaccurate. In this PR we generate all the fake data outside the loop to avoid this.

Other bug fixes:
- Only start polling GPU utilization after warmup event is complete
- Correct calculation of throughput: previously `(num_batches * batch_size) / sum(response_times)`, should have been `(num_batches * batch_size) / (last_response_time - first_request_time)`
- Make sure that response sent back to frontend is on CPU
- Use a semaphore to ensure writing to `metrics_dict` in `metrics_thread` and `gpu_utilization_thread` in a thread-safe manner




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix bugs in metrics calculation in inference benchmark and rerun baseline"

Before this PR, each `request_time` was separated by the time for a `torch.randn(...)` to create the fake `data` tensor on CPU. This meant that the gap between `request_times` **scaled with the batch_size**. So the latency comparisons across batch sizes were inaccurate. In this PR we generate all the fake data outside the loop to avoid this.

Other bug fixes:
- Only start polling GPU utilization after warmup event is complete
- Correct calculation of throughput: previously `(num_batches * batch_size) / sum(response_times)`, should have been `(num_batches * batch_size) / (last_response_time - first_request_time)`
- Make sure that response sent back to frontend is on CPU
- Use a semaphore to ensure writing to `metrics_dict` in `metrics_thread` and `gpu_utilization_thread` in a thread-safe manner




[ghstack-poisoned]

albanD commented on benchmarks/inference/server.py:
```diff
@@ -49,18 +60,21 @@ def _run_metrics(self):
         self.poll_gpu = False
 
         response_times = np.array(response_times)
+        metrics_lock.acquire()
```
nit you can do `with metrics_lock:`

albanD commented on benchmarks/inference/server.py:
```diff
@@ -87,38 +101,52 @@ def get_gpu_utilization():
-            time.sleep(0.1)
 
+        metrics_lock.acquire()
         self.metrics_dict["gpu_util"] = np.array(gpu_utilizations).mean()
```
nit: torch.tensor haha

albanD reviewed (approved):
Nice cleanup!

mikaylagawarecki committed:
Update on "Fix bugs in metrics calculation in inference benchmark and rerun baseline"

Before this PR, each `request_time` was separated by the time for a `torch.randn(...)` to create the fake `data` tensor on CPU. This meant that the gap between `request_times` **scaled with the batch_size**. So the latency comparisons across batch sizes were inaccurate. In this PR we generate all the fake data outside the loop to avoid this.

Other bug fixes:
- Only start polling GPU utilization after warmup event is complete
- Correct calculation of throughput: previously `(num_batches * batch_size) / sum(response_times)`, should have been `(num_batches * batch_size) / (last_response_time - first_request_time)`
- Make sure that response sent back to frontend is on CPU
- Use a lock to ensure writing to `metrics_dict` in `metrics_thread` and `gpu_utilization_thread` in a thread-safe manner




[ghstack-poisoned]
